lib/fastlane/plugin/wpmreleasetoolkit/helper/strings_file_validation_helper.rb

```diff
@@ -0,0 +1,129 @@
+module Fastlane
+  module Helper
+    class StringsFileValidationHelper
+      # context can be one of:
+      #   :root, :maybe_comment_start, :in_line_comment, :in_block_comment,
+      #   :maybe_block_comment_end, :in_quoted_key,
+      #   :after_quoted_key_before_eq, :after_quoted_key_and_equal,
+      #   :in_quoted_value, :after_quoted_value
+      State = Struct.new(:context, :buffer, :in_escaped_ctx, :found_key, keyword_init: true)
+
+      TRANSITIONS = {
+        root: {
+          /\s/ => :root,
+          '/' => :maybe_comment_start,
+          '"' => :in_quoted_key
+        },
+        maybe_comment_start: {
+          '/' => :in_line_comment,
+          /\*/ => :in_block_comment
+        },
+        in_line_comment: {
+          "\n" => :root,
+          /./ => :in_line_comment
+        },
+        in_block_comment: {
+          /\*/ => :maybe_block_comment_end,
+          /./m => :in_block_comment
+        },
+        maybe_block_comment_end: {
+          '/' => :root,
+          /./m => :in_block_comment
+        },
+        in_quoted_key: {
+          '"' => lambda do |state, _|
+            state.found_key = state.buffer.string.dup
+            state.buffer.string = ''
+            :after_quoted_key_before_eq
+          end,
+          /./ => lambda do |state, c|
+            state.buffer.write(c)
+            :in_quoted_key
+          end
+        },
+        after_quoted_key_before_eq: {
+          /\s/ => :after_quoted_key_before_eq,
+          '=' => :after_quoted_key_and_eq
+        },
+        after_quoted_key_and_eq: {
+          /\s/ => :after_quoted_key_and_eq,
+          '"' => :in_quoted_value
+        },
+        in_quoted_value: {
+          '"' => :after_quoted_value,
+          /./m => :in_quoted_value
+        },
+        after_quoted_value: {
+          /\s/ => :after_quoted_value,
+          ';' => :root
+        }
+      }.freeze
+
+      # Returns all the keys in the given `.strings` file, and the line(s) at
+      # which they are defined.
+      #
+      # @param [String] file The path to the file to inspect.
+      # @return [<Hash>] A `Hash` with keys the key values read from the
+      #         `.strings` file. The value of each entry is an array with the
+      #         line numbers where the key occurs.
+      def self.get_keys_from_strings(file:)
+        keys_with_lines = {}
+
+        state = State.new(context: :root, buffer: StringIO.new, in_escaped_ctx: false, found_key: nil)
+
+        File.readlines(file).each_with_index do |line, line_no|
+          line.chars.each_with_index do |c, col_no|
+            # Handle escaped characters at a global level. This is more
+            # straightforward than having a `TRANSITIONS` table that account
+            # for it.
```

Comment:
Not the first time I see you do hard-line-wrap instead of letting your IDE do the soft-wrapping depending on your IDE's preferences and line length settings 🤔 

Any strong reason for it? I mean, I know that it's gonna be akin to the old "tabs vs spaces" eternal debate asking this here 😅  but what surprises me most is that you always hard-wrap at 80 characters, while [even our current `rubocop` settings sets a way higher limit of 300](https://github.com/wordpress-mobile/release-toolkit/blob/trunk/.rubocop.yml#L60-L61) there.

I'm not completely against hard-wrapping as opposed to soft-wrapping… as long as it's consistent with the line length we have set ourselves for all the rest of the codebase (i.e. use the same limit for the comments as the one forced by rubocop for the code itself) 🙃 

Also (and this is mostly the main reason why I've noticed your 80-cols manual hard-wrapping in the first place), the fact that doing so leads to odd splits with comment lines sometimes only containing one or 2 words makes it a bit of an odd ball imho, as it increases vertical space those doc comments take just to account for 2 lone words 🤷 

Again, not a big deal, and very opinionated take from me here obviously, but since that's not the first time I see such lone words isolated on their own comment line _just_ to make that hard-wrap at 80-cols, I figured I'd still mention it 😃 . Personally I prefer more lose wrapping rules and using CRLFs for semantic breaks, e.g.:

```suggestion
            # Handle escaped characters at a global level.
            # This is more straightforward than having a `TRANSITIONS` table that account for it.
```

Reply:
> Any strong reason for it? I mean, I know that it's gonna be akin to the old "tabs vs spaces" eternal debate asking this here 😅 but what surprises me most is that you always hard-wrap at 80 characters

No _strong_ reason. My rationale for doing it is to avoid having to scroll horizontally when looking at diffs:

<img width="1491" alt="Screen Shot 2022-04-07 at 12 38 37 pm" src="https://user-images.githubusercontent.com/1218433/162111205-14162172-52b8-4d56-897a-6ce2e1a2439e.png">

I'm sure there's ways to introduce soft-wraps in most IDEs and diff tool. [GitHub _used_ to do it](https://github.blog/2013-12-03-soft-wrapping-on-prose-diffs/). Still, by enforcing short line in the source, I don't have to worry about it.

It obviously comes with tradeoffs. On top of what you already highlighted, it's also annoying to have to reselect and reformat comments every time they're changed because of that. But the result is net positive, IMHO.

Maybe this is a good chance to discuss the approach to take in this repo and document the choice?

Reply:
> [GitHub used to do it](https://github.blog/2013-12-03-soft-wrapping-on-prose-diffs/).

Interestingly, the suggestions render with soft wrap, but the diff in the PR `/files` do not 🤔

Reply:
> My rationale for doing it is to avoid having to scroll horizontally when looking at diffs:

I'm not sure if this is because they're currently A/B testing their new beta UI and I have it but your don't, or if it's a user setting in GitHub prefs that would make it differ between what you see in GitHub and what I see…

…but personally when I look at the diff in `/files` I have **plenty of space** left horizontally (largely enough room to fit waayyy more than 80 characters! 😛 ) before I'd need to scroll horizontally when viewing the diff on GitHub 🤔 

<img width="1899" alt="image" src="https://user-images.githubusercontent.com/216089/162291220-31da00e5-7e3a-46d4-9c40-ff69e79cdeed.png">

Reply:
Well if you put it like that...

<img width="3008" alt="Screen Shot 2022-04-11 at 1 26 18 pm" src="https://user-images.githubusercontent.com/1218433/162660295-977b4172-e434-43bc-9119-b80eb74c5dc9.png">

🙃

I formed this habit before MacBooks had Retina displays or I could afford and extra wide external monitor, and it's based on really [old conventions](https://stackoverflow.com/questions/2290016/git-commit-messages-50-72-formatting) from back in the day. Maybe it's time to revisit the length at which I make my editor hard-wrap...

What would be the "best" value for you @AliSoftware? I'll open a dedicated PR for that.

Reply:
> I formed this habit before MacBooks had Retina displays or I could afford and extra wide external monitor

Oh yes, and I do remember that time where I limited everything to 80 cols too and I had smaller monitors too 😛 But computer screen sizes have evolved since, and most importantly this length limit was for back in the days when Terminals were also 80 cols. Even if I were to run `bat` to list code in my Terminal, my terminal's preferences is set to use 240x60 by default, so… 🙃 

> based on really [old conventions](https://stackoverflow.com/questions/2290016/git-commit-messages-50-72-formatting) from back in the day

Btw for _commit messages_ I'm ok with trying to limit the length of the first line; but I'd not hard-wrap the following lines (aka the "summary"). So my rule here would be maybe something like 50/∞ instead of the 50/72 mentioned in your SO link 😛 

> What would be the "best" value for you @AliSoftware? I'll open a dedicated PR for that.

I have no strict limit to give you, as in practice:
 - I don't really set a hard limit at which I should hard-wrap
 - And instead I generally don't hard wrap (but I go to the next line to break long prose in paragraphs…)
 - Or when I do, I just use my gut feel to cut at a point in the very long lines (>200, 300?) where the hard-wrap would (1) not feel too odd in this part of the sentence, and (2) does not create imbalance with the rest of the line being wrapped. For example if a line is 310 characters long, I'll not hard-wrap it at 300 characters and leave the remaining 10 characters in the 2nd line, but instead maybe cut around 160/150? Or 120, 120, 70?

So basically I don't really have a hard rule myself… except that the rule that I don't have a hard rule 😛  and a hard line length limit that I'd follow blindly.
That being said, this repo is [currently configured to have a line length of 300 already for ruby code](https://github.com/wordpress-mobile/release-toolkit/blob/trunk/.rubocop.yml#L60-L61), so if you really want a number I'd go with that one for no particular reason other than that's what our config currently says (and what I believe rubocop suggested as a value back when we set it up based on its analysis of our current code base and max line lengths used throughout our existing code) 🙃